### PR TITLE
Added bookmark Wikidata items option

### LINKED ISSUE #3700: Bookmark Wikidata item
Currently, only pictures and locations can be bookmarked.
It would be useful if Wikidata items could be bookmarked too.
For instance if every month I go to the harbour and takes pictures of all kinds of warships, it would be useful to have a bookmark on item [warship](https://www.wikidata.org/wiki/Q3114762) so that I can easily access it and browse its subclasses to find the best items each of my pictures depict.

### PR DESCRIPTION
**Pull request by Ayan-10**
**Description**

- Option for bookmarking wikidata items. 
- New tab on Explore for viewing bookmarked wiki data

Fixes #3700

### CODE REVIEW

**Comment by Ayan-10** [comment]
@nicolas-raoul I solved the issue.

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
I am glad that you were able to fix the previous issue! :-)

I found a minor bug:
1) Tap "Explore"
2) Tap the search loop
3) Enter "rabbit"
4) Tap the "Items" tab
4) In the "Media" tab, tap any rabbit picture
5) In the media details activity that appears, tap the star
6) Press "Back"
7) Problem: The item now has its star enabled.

In this scenario, the star on the picture's media details page was tapped, not the star on the item page, so the picture only should have been bookmarked, not the item.

**Comment by Ayan-10** [comment]
@nicolas-raoul Actually for me no images are showing in Items -> Media. I uploaded some images and select depicts Rabbit. But even those images are not showing in the Items -> Media. Can you tell me how can I see images in Items -> Media?

**Comment by Ayan-10** [comment]
@nicolas-raoul I think you missed this [comment](https://github.com/commons-app/apps-android-commons/pull/4515#issuecomment-887194626).

**Comment by nicolas-raoul** [comment]
Here is what I see:
adb

**Comment by nicolas-raoul** [comment]
Sorry I had forgotten step 5 in my steps above. Can you try again? Thanks!

**Comment by Ayan-10** [comment]
@nicolas-raoul I can't see those images there. 

**Here is what I see for every item.
Latest master Pixel 2 API 30**



Can you tell me how can I see those images so that I can reproduce the issue and fix it?

**Comment by nicolas-raoul** [comment]
I think you have found a new bug to fix 🙂

Could you please create an issue about it, including a short screencast
from the app's home activity, and the logcat? Thanks!

**Comment by Ayan-10** [comment]
@nicolas-raoul I think I solved the problem. Please review it once.

**Comment by Ayan-10** [comment]
@nicolas-raoul Can you please review this PR? I made the requested changes on Aug 10.

**Comment by nicolas-raoul** [comment]
Sorry for the very late review!

I have found something that might be a bug:
1. Open the app
2. Select the `Explore` tab
3. In the search field, enter `plaque met 230711`
4. Tap the only media result that appears
5. Its media details show the depictions, tap its `bird` depiction
6. The `bird` Wikidata item is now shown. Problem: There is no star button. Is that intended?

adb

Thanks!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
How about some unit tests? They can probably be derived from the unit tests of the other types of bookmarks.

**Comment by Ayan-10** [comment]
> Sorry for the very late review!
> 
> I have found something that might be a bug:
> 
> 1. Open the app
> 2. Select the `Explore` tab
> 3. In the search field, enter `plaque met 230711`
> 4. Tap the only media result that appears
> 5. Its media details show the depictions, tap its `bird` depiction
> 6. The `bird` Wikidata item is now shown. Problem: There is no star button. Is that intended?
> 
> adb
> 
> Thanks!

@nicolas-raoul Yes that is intended, because when you are opening a depiction from a Media. It's calling WikiDataActivity from MediaDetailFragment and in MediaDetailFragment DepictItem is not provided, from MediaDetailActivity we only get the id and name of the depiction and If we put only name and id instead of a DepictedItem(Which contains more information than only name and string) it will behave improperly. That's why I intentionally removed the star button only when the WikiDataActivity is calling from MediaDetailActivity.

**Comment by nicolas-raoul** [comment]
It would be better to show the bookmark button always on WikiDataActivity, even if it has been opened from somewhere else. It sounds feasible even with the id only.

**Comment by Ayan-10** [comment]
@nicolas-raoul Yes, that is possible I just have to make a network call before calling WikiDataActivity from MediaDetailsFragment to get other details from the server. If there is no problem with that I will do that.

**Comment by nicolas-raoul** [comment]
Why not perform this network call only when the bookmark icon is tapped?
Or implement another bookmark creation method that takes as argument the Wikidata QID, performs the network call if necessary, then call the existing bookmark creation method.

**Comment by Ayan-10** [comment]
@nicolas-raoul Great idea. That can be done.

Thanks.

**Comment by Ayan-10** [comment]
@nicolas-raoul Now bookmark is accessible from MediaDetailFragment. Please review it. Also as clicking the icon performing a network call it's taking a little time to change the state of the star. Should I make a progress bar so that user can understand some task is going on?

**Review by nicolas-raoul — COMMENTED** [review]
Great, thanks!
Indeed it takes about a second. A UI to make the wait less worrying would be a great plus, but even like this I think it is not bad. I let you judge :-)

File: .idea/inspectionProfiles/Project_Default.xml [hunk 1]
```diff
@@ -2,15 +2,16 @@
   <profile version="1.0">
     <option name="myName" value="Project Default" />
     <inspection_tool class="AndroidLintNewerVersionAvailable" enabled="true" level="WARNING" enabled_by_default="true" />
-    <inspection_tool class="AutoCloseableResource" enabled="true" level="WARNING" enabled_by_default="true" />
+    <inspection_tool class="AutoCloseableResource" enabled="true" level="WARNING" enabled_by_default="true">
+      <option name="METHOD_MATCHER_CONFIG" value="java.util.Formatter,format,java.io.Writer,append,com.google.common.base.Preconditions,checkNotNull,org.hibernate.Session,close,java.io.PrintWriter,printf,android.database.sqlite.SQLiteOpenHelper,getWritableDatabase|getReadableDatabase,javax.inject.Provider,get" />
+    </inspection_tool>
```
Comment (nicolas-raoul) [inline .idea/inspectionProfiles/Project_Default.xml:7]: Are the changes to this file intended?

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsContentProvider.java [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+import android.text.TextUtils;
+import androidx.annotation.NonNull;
+import fr.free.nrw.commons.BuildConfig;
+import fr.free.nrw.commons.data.DBOpenHelper;
+import fr.free.nrw.commons.di.CommonsDaggerContentProvider;
+import javax.inject.Inject;
+import timber.log.Timber;
+
+/**
+ * Handles private storage for Bookmark pictures
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsContentProvider.java:20]: Bookmark pictures -> bookmarked items
I guess?

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsFragment.java [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+import butterknife.ButterKnife;
+import dagger.android.support.DaggerFragment;
+import fr.free.nrw.commons.R;
+import fr.free.nrw.commons.upload.structure.depictions.DepictedItem;
+import java.util.List;
+import javax.inject.Inject;
+import org.jetbrains.annotations.NotNull;
+
+/**
+ * Tab fragment to show list of Wikidata Items
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsFragment.java:25]: list of Wikidata Items -> list of bookmarked Wikidata items

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

File: .idea/inspectionProfiles/Project_Default.xml (hunk 1 above)
Comment (Ayan-10) [reply .idea/inspectionProfiles/Project_Default.xml]: No, I think those lines are added automatically.

File: .idea/inspectionProfiles/Project_Default.xml [hunk 4]
```diff
@@ -10,7 +10,6 @@
     <inspection_tool class="ControlFlowStatementWithoutBraces" enabled="true" level="ERROR" enabled_by_default="true" />
     <inspection_tool class="DefaultNotLastCaseInSwitch" enabled="true" level="WARNING" enabled_by_default="true" />
     <inspection_tool class="ExplicitThis" enabled="true" level="WEAK WARNING" enabled_by_default="true" />
-    <inspection_tool class="FieldMayBeFinal" enabled="true" level="WARNING" enabled_by_default="true" />
```
Comment (nicolas-raoul) [inline .idea/inspectionProfiles/Project_Default.xml:13]: Would you mind reverting this change too, except if there is a good reason for it? Thanks!

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
One minor change request.

Also, what about unit tests?
I guess at least the pictures bookmarks feature or place bookmarks feature have unit tests, right?

Otherwise the code looks good to me :-)
Thanks!

**Comment by Ayan-10** [comment]
> One minor change request.
> 
> Also, what about unit tests? I guess at least the pictures bookmarks feature or place bookmarks feature have unit tests, right?
> 
> Otherwise the code looks good to me :-) Thanks!

@nicolas-raoul I am on it. I will ping you as soon as I finish.

**Comment by Ayan-10** [comment]
@nicolas-raoul Tests are added, requested changes are made. Please review it.

File: app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsAdapter.kt [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        return BookmarkItemViewHolder(v)
+    }
+
+    override fun onBindViewHolder(holder: BookmarkItemViewHolder, position: Int) {
+
+        val depictedItem = list[position]
+        holder.depictsLabel.text = depictedItem.name
+        holder.description.text = depictedItem.description
+
+        if (depictedItem.imageUrl?.isNotBlank() == true) {
```
Comment (nicolas-raoul) [inline app/src/main/java/fr/free/nrw/commons/bookmarks/items/BookmarkItemsAdapter.kt:41]: Sorry I just spotted this. Could you please remove ` == true`?

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]

**Review by nicolas-raoul — CHANGES_REQUESTED** [review]
I found a crash:
1. Go to Explore
2. Search for some common word such as "street"
3. Bookmark the 10 first Wikidata items
4. Go back, then to the Bookmarks tab
5. Open the Items tab
6. Crash:
```
java.lang.NullPointerException
at fr.free.nrw.commons.bookmarks.items.BookmarkItemsAdapter.onBindViewHolder(BookmarkItemsAdapter.kt:41)
at fr.free.nrw.commons.bookmarks.items.BookmarkItemsAdapter.onBindViewHolder(BookmarkItemsAdapter.kt:18)
at androidx.recyclerview.widget.RecyclerView$Adapter.onBindViewHolder(RecyclerView.java:7093)
at androidx.recyclerview.widget.RecyclerView$Adapter.bindViewHolder(RecyclerView.java:7172)
at androidx.recyclerview.widget.RecyclerView$Recycler.tryBindViewHolderByDeadline(RecyclerView.java:6037)
at androidx.recyclerview.widget.RecyclerView$Recycler.tryGetViewHolderForPositionByDeadline(RecyclerView.java:6304)
at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6143)
at androidx.recyclerview.widget.RecyclerView$Recycler.getViewForPosition(RecyclerView.java:6139)
at androidx.recyclerview.widget.LinearLayoutManager$LayoutState.next(LinearLayoutManager.java:2314)
at androidx.recyclerview.widget.LinearLayoutManager.layoutChunk(LinearLayoutManager.java:1631)
at androidx.recyclerview.widget.LinearLayoutManager.fill(LinearLayoutManager.java:1591)
at androidx.recyclerview.widget.LinearLayoutManager.onLayoutChildren(LinearLayoutManager.java:668)
at androidx.recyclerview.widget.RecyclerView.dispatchLayoutStep2(RecyclerView.java:4141)
at androidx.recyclerview.widget.RecyclerView.dispatchLayout(RecyclerView.java:3853)
at androidx.recyclerview.widget.RecyclerView.onLayout(RecyclerView.java:4410)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at androidx.viewpager.widget.ViewPager.onLayout(ViewPager.java:1775)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.RelativeLayout.onLayout(RelativeLayout.java:1083)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1812)
at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1656)
at android.widget.LinearLayout.onLayout(LinearLayout.java:1565)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.widget.FrameLayout.layoutChildren(FrameLayout.java:323)
at android.widget.FrameLayout.onLayout(FrameLayout.java:261)
at com.android.internal.policy.DecorView.onLayout(DecorView.java:1041)
at android.view.View.layout(View.java:22420)
at android.view.ViewGroup.layout(ViewGroup.java:6579)
at android.view.ViewRootImpl.performLayout(ViewRootImpl.java:3343)
at android.view.ViewRootImpl.performTraversals(ViewRootImpl.java:2807)
at android.view.ViewRootImpl.doTraversal(ViewRootImpl.java:1853)
at android.view.ViewRootImpl$TraversalRunnable.run(ViewRootImpl.java:8476)
at android.view.Choreographer$CallbackRecord.run(Choreographer.java:949)
at android.view.Choreographer.doCallbacks(Choreographer.java:761)
at android.view.Choreographer.doFrame(Choreographer.java:696)
at android.view.Choreographer$FrameDisplayEventReceiver.run(Choreographer.java:935)
at android.os.Handler.handleCallback(Handler.java:873)
at android.os.Handler.dispatchMessage(Handler.java:99)
at android.os.Looper.loop(Looper.java:214)
at android.app.ActivityThread.main(ActivityThread.java:7032)
at java.lang.reflect.Method.invoke(Native Method)
at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:494)
at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:965)
```

**Comment by Ayan-10** [comment]
@nicolas-raoul It was happening because of removing "== true". If there is no specific problem can I keep "== true".

**Comment by nicolas-raoul** [comment]
Very sorry I had no idea this could have this kind of effect!
Yes if that fixes the problem that's great. Thanks!

**Review by nicolas-raoul — APPROVED** [review]

**Comment by nicolas-raoul** [comment]
That indeed was the problem, `== true` is an anti-pattern in Java but in Kotlin it is actually useful, I learnt something today :-)

Thanks a lot for this useful feature!

**Comment by Ayan-10** [comment]
@nicolas-raoul Actually `isNotBlank()` in kotlin returns ` Boolean? ` instead of `Boolean` that is why we just can not remove `== true` (As `== true` only work on Boolean). if we want to remove it we have to put the not-null assertion operator (!!) and when the imageUrl is null because of the not-null assertion operator (!!) we are getting NPE. In java, we can just remove `==true`. As the variable can only return `Boolean`.

Thanks for your appriciation.

**Pull request merged**